### DISCUSSION TITLE
Feature Idea: auto normalisation-gain-type

### DISCUSSION THREAD

**Discussion by JasonLG1979** [librespot-org/librespot#780] [Feature Requests]
Librespot and Spotify currently supports both track and album normalisation gain types. Which to use is currently up to the user, defaulting to album. An auto mode would not be that difficult. If 2 tracks in a row are from the same album use album mode if not use track mode.

**Comment by JasonLG1979** [comment]
Basically if the current track is the same album as the next track, album gain if not track gain. The only special case is at the end of the playlist. In that case since there would be no next track you'd want to look at the previous track to see if it's the same album to decide.

**Comment by JasonLG1979** [comment]
My bad. Just noticed the whole feature requests should be discussions thing...

**Comment by roderickvd** [comment]
I can look into it but want to get a `cpal` backend done first.

**Reply by JasonLG1979** [reply]
Absolutely no rush. It's just a wishlist feature. The music I listen to is mostly loudness war victims so what I've seen when paying attention to librespot debug output is that track and album gain generally aren't that far off from each other. So for me it may not make that huge of a difference. It also wouldn't make that much of a difference for someone who either always listens to mixed playlists or someone that always listens whole albums at a time. Currently they can just pick the appropriate mode. It would make for a good catch all default though I think that would make everyone happy. Because I think that most people are probably somewhere in the middle of those 2 extreme use cases.      

In general what do you think of it as an idea? And also does the (very) rough implementation sketch of how it should work make sense to you? Or can you think of a better way it should work?

**Reply by roderickvd** [reply]
I like the idea and as a user, think that knob should be smarter too. And it’s what the official Spotify player does.

I’m still finding my way in the “Connect” parts of librespot, I have to investigate what it would take to pass a “shuffle or album mode” to the player where the normalization takes place.

**Reply by JasonLG1979** [reply]
> I like the idea and as a user, think that knob should be smarter too. And it’s what the official Spotify player does.

Good deal. I have no idea what an official client does. That just seemed like the most simple way to make it "smarter". 

> I’m still finding my way in the “Connect” parts of librespot, I have to investigate what it would take to pass a “shuffle or album mode” to the player where the normalization takes place.

Do you mean track mode instead of shuffle?

If you mean track mode you got me that's over my head,lol!!!

But if you mean shuffle as in shuffle or repeat the question then becomes is shuffle a client side thing, as in is the client given a playlist and then told to randomize it? And how do you treat repeat?

**Reply by roderickvd** [reply]
> Do you mean track mode instead of shuffle?

Yes that is what I meant. Quoting https://artists.spotify.com/help/article/loudness-normalization:

> 1. We normalize an entire album at the same time, so gain compensation doesn’t change between tracks. This means the softer tracks are as soft as you intend them to be.
> 2. We adjust individual tracks when shuffling an album or listening to tracks from multiple albums (e.g. listening to a playlist).

-- 

> But if you mean shuffle as in shuffle or repeat the question then becomes is shuffle a client side thing, as in is the client given a playlist and then told to randomize it? And how do you treat repeat?

I don't know yet, but I believe repeat doesn't work currently. So I suspect that it is server-side. I've seen a couple of related commits in `librespot-java`. Hope to discover and fix this in `new-api`.

**Comment by roderickvd** [comment]
I created PR #844 for this one, let's continue discussions there.
